rust/src/delta.rs

```diff
@@ -426,6 +426,49 @@ impl DeltaTableState {
     pub fn current_metadata(&self) -> Option<&DeltaTableMetaData> {
         self.current_metadata.as_ref()
     }
+
+    /// merges new state information into our state
+    pub fn merge(&mut self, new_state: &mut DeltaTableState) {
+        // build a lookup table of all remove actions
+
+        if !new_state.tombstones.is_empty() {
+            let new_removals: HashSet<&str> = new_state
+                .tombstones
+                .iter()
+                .map(|s| s.path.as_str())
+                .collect();
+
+            // previous state files must filtered for removals,
+            // new state files are appended without looking.
+            self.files
+                .retain(|a| !new_removals.contains(a.path.as_str()));
```

Comment:
@dispanser we need to apply retain for new file list in new_state too if it is constructed from incremental updates instead of checkpoints right?

Reply:
Interesting question. In the current impleentation, `merge` is called by `apply_log_from_bufread`, which is in turn called from `apply_log` for each version separately. So if we assume that no add + remove for the same file happens in the same version, we're safe.

We may be better of doing that "merge" once for multiple commits, and in that case the curent implementation becomes wrong.

I think I'll change it as you suggest because the way it is currently implemented makesa bit of a hidden assumption about how it's called, and the generic `merge` operation should handle all scenarios.

Reply:
oh yeah, you are right, this is not a problem for the current PR then :+1: we can revisit this when we start doing parallel table load.

Reply:
I changed it to do the append first, which makes the code shorter because it allows me to remove the comments explaining the pecularities of the implementation :-).

One final thought wrt to this topic: we could argue that a `DeltaTableState` must be consistent / self-contained according to some rules, i.e. that "removals for its own files are already resolved". In that case we wouldn't need that.

Reply:
yeah, I think that's a good idea. It also avoids having to go through the files field twice, once for append, once for retain.